### PR TITLE
ARROW-2351 [C++] StringBuilder::append(vector<string>...) not impleme…

### PR DESCRIPTION
**Pull request by gaolizhou**
changed the API from`Status Append(const std::vector<std::string>& values, uint8_t* null_bytes);` to  `Status Append(const std::vector<std::string>& values);` IMO, if string is empty, then it should be null, and vice versa.

**[update]** change the API back to original.

### CODE REVIEW

**Comment by wesm** [comment]
> IMO, if string is empty, then it should be null, and vice versa.

We distinguish between empty strings and null strings

**Comment by gaolizhou** [comment]
> We distinguish between empty strings and null strings

hmmm, is it possible that  the string is not empty and the null bit is true ?

**Comment by xhochy** [comment]
> hmmm, is it possible that the string is not empty and the null bit is true ?

No, this should not happen.

**Comment by rvernica** [comment]
Do we want to update the API as well to `Status Append(const std::vector<std::string>& values, vector<bool> null_bytes);` to match the API for `NumericBuilder`?

**Comment by xhochy** [comment]
> Do we want to update the API as well to Status Append(const std::vector<std::string>& values, vector<bool> null_bytes); to match the API for NumericBuilder?

Usability-wise this definitely makes sense but this can also be done in a followup-PR.

**Comment by gaolizhou** [comment]
Please help review it  and leave your comments.

**Comment by gaolizhou** [comment]
can someone help review the code and leave your comments?
Thanks!

**Comment by cpcloud** [comment]
I don't think we can make a change like this. You need to be able to append a non null empty string. The fact that an empty string is what is at that memory location is an implementation detail.

**Comment by cpcloud** [comment]
Ah I see you've updated to keep the original API.

File: cpp/src/arrow/builder.cc [hunk 1]
```diff
@@ -1385,6 +1385,28 @@ const uint8_t* BinaryBuilder::GetValue(int64_t i, int32_t* out_length) const {
 
 StringBuilder::StringBuilder(MemoryPool* pool) : BinaryBuilder(utf8(), pool) {}
 
+Status StringBuilder::Append(const std::vector<std::string>& values,
+                             uint8_t* null_bytes) {
+  std::size_t total_length = std::accumulate(
+      values.begin(), values.end(), 0ULL,
+      [](uint64_t sum, const std::string& str) { return sum + str.size(); });
```
Comment (wesm) [inline cpp/src/arrow/builder.cc]: Is using `std::accumulate` with a lambda the same speed as a for loop?

**Review by wesm — COMMENTED** [review]
LGTM except for the lambda perf question

File: cpp/src/arrow/builder.cc (hunk 1 above)
Comment (gaolizhou) [reply cpp/src/arrow/builder.cc]: Tested with below code, for-loop-solution and accumulate-solution almost has same performance.To count 100000000 strings, for-loop-solution takes 21.2312 ms,
 accumulate-solution takes 21.5120 ms.

log:
```
lizgao@lizgao-ubuntu:~/CLionProjects/cmake-test$ ./cmake_test 
WARNING: Logging before InitGoogleLogging() is written to STDERR
I0330 10:29:56.864248  3989 accmulate-lambda-performance.cc:15] Begin for_loop_solution
I0330 10:29:57.076560  3989 accmulate-lambda-performance.cc:19] End for_loop_solution = 8000000000
I0330 10:29:57.076603  3989 accmulate-lambda-performance.cc:23] Begin accumulate_solution
I0330 10:29:57.291723  3989 accmulate-lambda-performance.cc:27] End accumulate_solution = 8000000000

```

```
//
// Created by lizgao on 3/30/18.
//
#include <cstdint>
#include <numeric>
#include <string>
#include <vector>
#include <glog/logging.h>

namespace {

void for_loop_solution(const std::vector<std::string> &values) {
  std::size_t total_length = 0;
  std::size_t cnt = values.size();
  LOG(INFO) << "Begin for_loop_solution";
  for (std::size_t i = 0; i < cnt; ++i) {
    total_length += values[i].size();
  }
  LOG(INFO) << "End for_loop_solution = " << total_length;
}

void accumulate_solution(const std::vector<std::string> &values) {
  LOG(INFO) << "Begin accumulate_solution";
  std::size_t total_length = std::accumulate(
        values.begin(), values.end(), 0ULL,
        [](uint64_t sum, const std::string& str) { return sum + str.size(); });
  LOG(INFO) << "End accumulate_solution = " << total_length;
}
}

void accumulate_lambda_performance_test() {
  std::vector<std::string> str_list(100000000, std::string(80, '*'));
  for_loop_solution(str_list);
  accumulate_solution(str_list);
}
```

**Comment by gaolizhou** [comment]
Is there any more comments?

**Review by xhochy — APPROVED** [review]
+1, LGTM

**Comment by xhochy** [comment]
@gaolizhou Thanks for implementing this and going through the review comments.

Do you have an Apache JIRA ID so I can assign https://issues.apache.org/jira/browse/ARROW-2351 to you?

**Comment by gaolizhou** [comment]
@xhochy  my JIRA Id is gaolizhou

**Comment by kou** [comment]
Sorry for my late comment.
It may be better that we use `const uint8_t* valid_bytes` instead of `uint8_t* null_bytes`. Because other `Append` API use `const uint8_t* valid_bytes`.

**Comment by kou** [comment]
I opened a JIRA issue https://issues.apache.org/jira/browse/ARROW-2388 and a pull request #1833.

**Pull request closed without merging**